[CI] IpFieldBlockLoaderTests testBlockLoaderForFieldInObject {preference=Params[syntheticSource=true, preference=STORED]} failing

**Build Scans:**
- [elasticsearch-periodic #10127 / openjdk24_checkpart1_java-matrix](https://gradle-enterprise.elastic.co/s/txm6gom2tzhum)
- [elasticsearch-periodic-platform-support #10130 / almalinux-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/nfkvb722lrmaq)
- [elasticsearch-periodic-platform-support #10130 / debian-12_platform-support-unix](https://gradle-enterprise.elastic.co/s/6i6uolted5c4e)
- [elasticsearch-pull-request #86726 / part-1](https://gradle-enterprise.elastic.co/s/vurej5luhg2as)
- [elasticsearch-pull-request #86586 / part-1](https://gradle-enterprise.elastic.co/s/e3siufncjjnhe)
- [elasticsearch-pull-request #86552 / part-1](https://gradle-enterprise.elastic.co/s/2nmzs4hddyg4c)
- [elasticsearch-pull-request #86536 / part-1](https://gradle-enterprise.elastic.co/s/xr2umi374kkuk)
- [elasticsearch-pull-request #86536 / part-1](https://gradle-enterprise.elastic.co/s/n6nxdvr7rs2xs)
- [elasticsearch-pull-request #86496 / part-1](https://gradle-enterprise.elastic.co/s/n75h73cpghl22)
- [elasticsearch-pull-request #86496 / part-1](https://gradle-enterprise.elastic.co/s/po7pbuikwo4ni)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.index.mapper.blockloader.IpFieldBlockLoaderTests.testBlockLoaderForFieldInObject {preference=Params[syntheticSource=true, preference=STORED]}" -Dtests.seed=9D5C35625B78BBA5 -Dtests.locale=ps-AF -Dtests.timezone=Pacific/Gambier -Druntime.java=24
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.index.mapper.blockloader.IpFieldBlockLoaderTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testBlockLoaderForFieldInObject%20%7Bpreference%3DParams%5BsyntheticSource%3Dtrue%2C%20preference%3DSTORED%5D%7D'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: <[[0 0 0 0 0 0 0 0 0 0 ff ff c4 5c 91 5f], [34 95 db c0 ff ff ff ff b1 a4 9b bb ff ff ff ff], [ea 4a c6 60 0 0 0 0 c3 e3 25 5e 0 0 0 0]]>
     but: was <[[0 0 0 0 0 0 0 0 0 0 ff ff c4 5c 91 5f], [34 95 db c0 ff ff ff ff b1 a4 9b bb ff ff ff ff], [ea 4a c6 60 0 0 0 0 0 0 c3 e3 25 5e 0 0]]>
```

**Issue Reasons:**
- [main] 15 failures in test testBlockLoaderForFieldInObject {preference=Params[syntheticSource=true, preference=STORED]} (4.7% fail rate in 322 executions)
- [main] 12 failures in step part-1 (6.9% fail rate in 174 executions)
- [main] 6 failures in pipeline elasticsearch-pull-request (3.6% fail rate in 168 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -576,24 +576,6 @@ tests:
 - class: org.elasticsearch.xpack.ml.integration.TextEmbeddingQueryIT
   method: testModelWithPrefixStrings
   issue: https://github.com/elastic/elasticsearch/issues/133138
-- class: org.elasticsearch.index.mapper.blockloader.IpFieldBlockLoaderTests
-  method: testBlockLoaderOfMultiField {preference=Params[syntheticSource=true, preference=DOC_VALUES]}
-  issue: https://github.com/elastic/elasticsearch/issues/133149
-- class: org.elasticsearch.index.mapper.blockloader.IpFieldBlockLoaderTests
-  method: testBlockLoaderOfMultiField {preference=Params[syntheticSource=true, preference=STORED]}
-  issue: https://github.com/elastic/elasticsearch/issues/133150
-- class: org.elasticsearch.index.mapper.blockloader.IpFieldBlockLoaderTests
-  method: testBlockLoaderOfMultiField {preference=Params[syntheticSource=true, preference=NONE]}
-  issue: https://github.com/elastic/elasticsearch/issues/133151
-- class: org.elasticsearch.index.mapper.blockloader.IpFieldBlockLoaderTests
-  method: testBlockLoaderForFieldInObject {preference=Params[syntheticSource=true, preference=STORED]}
-  issue: https://github.com/elastic/elasticsearch/issues/133178
-- class: org.elasticsearch.index.mapper.blockloader.IpFieldBlockLoaderTests
-  method: testBlockLoaderForFieldInObject {preference=Params[syntheticSource=true, preference=DOC_VALUES]}
-  issue: https://github.com/elastic/elasticsearch/issues/133179
-- class: org.elasticsearch.index.mapper.blockloader.IpFieldBlockLoaderTests
-  method: testBlockLoaderForFieldInObject {preference=Params[syntheticSource=true, preference=NONE]}
-  issue: https://github.com/elastic/elasticsearch/issues/133180
 - class: org.elasticsearch.test.rest.yaml.CcsCommonYamlTestSuiteIT
   method: test {p0=search.vectors/90_sparse_vector/Indexing and searching multi-value sparse vectors in >=8.15}
   issue: https://github.com/elastic/elasticsearch/issues/133184

--- a/server/src/main/java/org/elasticsearch/common/network/InetAddresses.java
+++ b/server/src/main/java/org/elasticsearch/common/network/InetAddresses.java
@@ -239,8 +239,8 @@ public class InetAddresses {
                 return null; // Invalid character
             }
         }
-        if (currentHextetStart != length) {
-            // Handle the last hextet
+        // Handle the last hextet unless the IP address ends with ::
+        if (currentHextetStart < offset + length) {
             if (putHextet(bytes, currentHextet) == false) {
                 return null;
             }

--- a/server/src/test/java/org/elasticsearch/common/network/InetAddressesTests.java
+++ b/server/src/test/java/org/elasticsearch/common/network/InetAddressesTests.java
@@ -27,6 +27,7 @@ import java.net.NetworkInterface;
 import java.net.UnknownHostException;
 import java.nio.charset.StandardCharsets;
 import java.util.Enumeration;
+import java.util.List;
 import java.util.Locale;
 
 import static org.hamcrest.Matchers.equalTo;
@@ -180,10 +181,12 @@ public class InetAddressesTests extends ESTestCase {
             // Shouldn't hit DNS, because it's an IP string literal.
             InetAddress ipv6Addr = InetAddress.getByName(ipString);
             assertEquals(ipv6Addr, InetAddresses.forString(ipString));
+            int extraLength = randomInt(8);
+            int offset = randomInt(extraLength);
             byte[] asBytes = ipString.getBytes(StandardCharsets.UTF_8);
-            byte[] bytes = new byte[32];
-            System.arraycopy(asBytes, 0, bytes, 8, asBytes.length);
-            assertEquals(ipv6Addr, InetAddresses.forString(bytes, 8, asBytes.length));
+            byte[] bytes = new byte[asBytes.length + extraLength];
+            System.arraycopy(asBytes, 0, bytes, offset, asBytes.length);
+            assertEquals(ipv6Addr, InetAddresses.forString(bytes, offset, asBytes.length));
             assertTrue(InetAddresses.isInetAddress(ipString));
         }
     }
@@ -206,6 +209,35 @@ public class InetAddressesTests extends ESTestCase {
         assertEquals("::1", InetAddresses.toAddrString(InetAddresses.forString("0:0:0:0:0:0:0:1")));
         assertEquals("2001:658:22a:cafe::", InetAddresses.toAddrString(InetAddresses.forString("2001:0658:022a:cafe::")));
         assertEquals("::102:304", InetAddresses.toAddrString(InetAddresses.forString("::1.2.3.4")));
+        assertEquals("2001:db8::1:0:0:1", InetAddresses.toAddrString(InetAddresses.forString("2001:db8::1:0:0:1")));
+    }
+
+    public void testWithOffsets() {
+        List<String> ipStrings = List.of(
+            "1:2:3:4:5:6:7:8",
+            "2001:0:0:4::8",
+            "2001::4:5:6:7:8",
+            "2001:0:3:4:5:6:7:8",
+            "0:0:3::ffff",
+            "::4:0:0:0:ffff",
+            "::5:0:0:ffff",
+            "1::4:0:0:7:8",
+            "::",
+            "::1",
+            "2001:658:22a:cafe::",
+            "::102:304",
+            "2001:db8::1:0:0:1",
+            "2001:db8::1:0:1:0"
+        );
+        for (String ipString : ipStrings) {
+            byte[] bytes = ipString.getBytes(StandardCharsets.UTF_8);
+            int extraLength = randomInt(8);
+            int offset = randomInt(extraLength);
+            byte[] bytesWithPadding = new byte[bytes.length + extraLength];
+            System.arraycopy(bytes, 0, bytesWithPadding, offset, bytes.length);
+            InetAddress addr = InetAddresses.forString(bytesWithPadding, offset, bytes.length);
+            assertEquals(ipString, InetAddresses.toAddrString(addr));
+        }
     }
 
     public void testToUriStringIPv4() {

